### PR TITLE
limit concurrency of fetching docs

### PR DESCRIPTION
**Pull request by PSeitz**
Limits the concurrency of fetching docs to 10

Addresses #1449

### CODE REVIEW

File: quickwit-search/src/fetch_docs.rs [hunk 1]
```diff
@@ -46,16 +47,16 @@ async fn fetch_docs_to_map<'a>(
         .collect();
 
     // We sort global hit addrs in order to allow for the grouby.
-    global_doc_addrs.sort_by_key(|global_doc_addr| global_doc_addr.split);
```
Comment (fulmicoton) [inline quickwit-search/src/fetch_docs.rs:49]: I'm lost. Why is it better than sort_by_key?

**Comment by PSeitz** [comment]
Removed reference from GlobalDocAddress, since I got a weird compiler error (compiler bug?) otherwise

```
error: implementation of `FnOnce` is not general enough
   --> quickwit-search/src/service.rs:151:43
    |
151 |       ) -> crate::Result<FetchDocsResponse> {
    |  ___________________________________________^
152 | |         let storage = self
153 | |             .storage_uri_resolver
154 | |             .resolve(&fetch_docs_request.index_uri)?;
...   |
163 | |         Ok(fetch_docs_response)
164 | |     }
    | |_____^ implementation of `FnOnce` is not general enough
    |
    = note: closure with signature `fn(GlobalDocAddress<'0>) -> impl futures::Future<Output = [async output]>` must implement `FnOnce<(GlobalDocAddress<'1>,)>`, for any two lifetimes `'0` and `'1`...
    = note: ...but it actually implements `FnOnce<(GlobalDocAddress<'_>,)>`
```

**Comment by fulmicoton** [comment]
@PSeitz Can you confirm having tested it on a large number of docs?

**Comment by PSeitz** [comment]
> @PSeitz Can you confirm having tested it on a large number of docs?

Yes, I can confirm it works

**Comment by fulmicoton** [comment]
@PSeitz can you confirm that at no point we debounce read calls to docs from a same docstore block, and open an issue accordingly? This could have massive performance gain for @kstaken use case described in #1449

**Review by fulmicoton — APPROVED** [review]

**Comment by fulmicoton** [comment]
@PSeitz 

> With this change we still may load the same block multiple times, but it's more unlikely. The current state is
> 10 concurrent doc requests (no optimization on same blocks), a cache of 100 and e.g. 10_000 docs.

Are you talking about tantivy's docstore cache?
Is this something you actually observed? I think we end up loading the same block most of the time.

I don't think there is any debouncing logic anywhere.
Let's assume we are loading 10 docs from the same block.
In that case, we do the fetch request all at the same time, which result in a cache miss and all request are emitted.


Let's assume we do 10k requests in random block, with big overlaps.
In that case the cache, the cache is not big enough to deal with this and we experience mostly cache misses.
We probably want to debounce requests at the storage layer.

**Comment by PSeitz** [comment]
> Are you talking about tantivy's docstore cache?
Is this something you actually observed? I think we end up loading the same block most of the time.

Yes tantivy's docstore cache. I did not observe it.

> I don't think there is any debouncing logic anywhere.
Let's assume we are loading 10 docs from the same block.
In that case, we do the fetch request all at the same time, which result in a cache miss and all request are emitted.

For TOP 10 docs it will have no effect, since they are executed concurrently.

> Let's assume we do 10k requests in random block, with big overlaps.
In that case the cache, the cache is not big enough to deal with this and we experience mostly cache misses.
We probably want to debounce requests at the storage layer.

Yes, but also depends on how many docs are there per block.

I was thinking about collecting the block ranges from the docs and then load them only once. 
On the storage layer it would also be possible with a cache or some timeout collector on the same byte range. This could have some synergies, but we could since have similar issues with cache invalidation.

**Pull request merged**